### PR TITLE
Fix mac tests with conda

### LINKED ISSUE #4939: macOS tests seem broken
I just noticed that macOS tests seem to be broken and effectively disabled, but they are still reporting success!

https://github.com/gwastro/pycbc/actions/runs/11835372970/job/32978009917

This problem might have started with #4922:

https://github.com/gwastro/pycbc/actions/runs/11669573223/job/32491822034?pr=4922

image

### PR DESCRIPTION
**Pull request by duncanmmacleod**
## Standard information about the request

This is a: bug fix

This change affects: continuous integration

This change changes: continuous integration

This change: has appropriate unit tests, follows style guidelines (See e.g. [PEP8](https://peps.python.org/pep-0008/)), has been proposed using the [contribution guidelines](https://github.com/gwastro/pycbc/blob/master/CONTRIBUTING.md)

This change will: n/a

## Motivation

Fixes #4939

## Contents

This PR fixes #4939 by updating the CI/tox configuration for macOS.

## Links to any issues or associated PRs

Fixes #4939 
Closes #4944 

## Additional notes

- [x] The author of this pull request confirms they will adhere to the [code of conduct](https://github.com/gwastro/pycbc/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
@titodalcanton @ahnitz, I think this fixes the test configuration, but exposes a real-world test failure:

```
E       AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?
```

I don't think this object is available on arm64. Can/should the calling code be patched to use `numpy.longdouble`?

**Comment by spxiwh** [comment]
The lapack pinned version was important ...

**Comment by duncanmmacleod** [comment]
> The lapack pinned version was important ...

Why? It is (I think) precisely the reason why the build is failing, that old version of lapack has not been, and will not be, built for macOS on ARM64.

**Comment by spxiwh** [comment]
@duncanmmacleod I think lapack stopped shipping the headers after this release, making it very hard to build codes from source that compile against it. There's also this weirdness about liblapacke, which I think also isn't in the new releases.

https://github.com/mikekatz04/BBHx?tab=readme-ov-file#prerequisites (because PyCBC does not use lapack directly, we use it via numpy like most).

I have built a BBHx env on my ARM mac though using this version of lapack ... Though it does cause an issue on the IGWN envs (where lapack, or the lapack headers, are not present), which is a discussion I'd been meaning to have with you at some point.

**Comment by spxiwh** [comment]
Although I now see that my mac env has lapack=3.9.0 installed, so I'm not sure how that worked/is working.

File: tox.ini [hunk 1]
```diff
@@ -28,20 +30,19 @@ deps =
...
+    c-compiler
+    cxx-compiler
     gsl
-    lapack==3.6.1
-conda_channels=conda-forge
-setenv =
-    ; Tell the linker to look for shared libs inside the temporary Conda env.
-    ; Needed to build BBHx's wheel, whick links to LAPACK.
-    LIBRARY_PATH={envdir}/lib:{env:LIBRARY_PATH:}
+    ligo-segments
```
Comment (ahnitz) [inline tox.ini:37]: A couple of these are python packages that we should be picking up through the requirements. At least

ligo-segments and python-ligo-lw should probably be removed from conda_deps, no?

Comment (ahnitz) [inline tox.ini:37]: @duncanmmacleod Thanks for looking into this!

A couple of these are python packages that we should be picking up through the requirements. At least

ligo-segments and python-ligo-lw should probably be removed from conda_deps, no?

Comment (duncanmmacleod) [reply tox.ini:37]: @ahnitz, these two packages are specifically called out in the conda requirements because PyPI only has a source distribution and these projects don't build cleanly with newer Python versions _and_ newer versions of Clang. This means that `pip install python-ligo-lw` fails.

The conda packages include the necessary patches to build, but that hasn't (yet) been backported into the source distribution.

**Comment by duncanmmacleod** [comment]
> Although I now see that my mac env has lapack=3.9.0 installed, so I'm not sure how that worked/is working.

This is the difference between the packages that provided the headers, and those that provide the libraries. The headers do seem to be a little muddled, it took me a while to find that we need the packages `liblapacke` to get `lapacke.h` and `openblas` to get `lapack.h`.

File: tox.ini (hunk 1 above)
Comment (ahnitz) [reply tox.ini:37]: @duncanmmacleod Ok, that's fair enough. In which case, I'd just request a comment explaining this. In principle these lines should be removed once those changes have been included in the upstream code.

Comment (duncanmmacleod) [reply tox.ini]: @ahnitz, I have separated things out so that there is a unique commit that adds these packages to the `tox.ini` file (with a comment) that can be simply reverted when appropriate.

**Comment by spxiwh** [comment]
> @titodalcanton @ahnitz, I think this fixes the test configuration, but exposes a real-world test failure:
> 
> ```
> E       AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?
> ```
> 
> I don't think this object is available on arm64. Can/should the calling code be patched to use `numpy.longdouble`?

@duncanmmacleod The failure related to this that I see above is in a test case that is really designed to test "have we broken anything in PyCBC Live". It compares to an old version of PyCBC Live which wouldn't run on a mac (though the test may now run on a mac with this change), so this comparison may not make sense on a mac ... But all other places have changed `float128` to `numpy.longdouble` ... I'm confused how this didn't break before??

So in short, I would be happy with just changing this. If the test still fails, disable it on mac.

**Comment by duncanmmacleod** [comment]
> So in short, I would be happy with just changing this. If the test still fails, disable it on mac.

@spxiwh, I took the liberty of adding that change, and it fixes the tests, so I'm happy.

There are _other_ uses of `numpy.float128`, e.g. https://github.com/gwastro/pycbc/blob/161bcea4a608a370fb06bd6e6d81b77ce0d03638/pycbc/waveform/pycbc_phenomC_tmplt.py#L21, but I am making no attempt to address those.

**Comment by spxiwh** [comment]
Interesting! I hadn't been aware that module even existed. Something to be removed, I think!

File: tox.ini [hunk 2]
```diff
@@ -73,21 +75,20 @@ deps =
...
-    lin: gxx_linux-64>=12.2.0
-    ; mac doesn't work, need fix
-    ; mac: clang_osx-64
-    ; mac: clangxx_osx-64
-    binutils_linux-64>=2.39
-    gsl
-    lapack==3.6.1
-    openmpi
-conda_channels=conda-forge
+    mysqlclient
```
Comment (titodalcanton) [inline tox.ini:82]: Duplicated line?

**Comment by titodalcanton** [comment]
I think this also cleans up a bit the problem mentioned in #4857. Thanks @duncanmmacleod.

File: tox.ini (hunk 2 above)
Comment (duncanmmacleod) [reply tox.ini]: Good catch, this has been removed.

File: .github/workflows/mac-test.yml [hunk 3]
```diff
@@ -13,16 +13,46 @@ jobs:
...
+
+    - name: Conda info
+      run: conda info --all
+
+    - name: Install tox
+      run: |
+        conda install \
+            pip \
+            setuptools \
+            "tox<4.0.0"
```
Comment (duncanmmacleod) [inline .github/workflows/mac-test.yml:54]: Is this restriction understood?

**Comment by titodalcanton** [comment]
We list the same `conda-deps` twice now, is it possible to deduplicate that with something like `{[base-conda-deps]}`?

File: .github/workflows/mac-test.yml (hunk 3 above)
Comment (spxiwh) [reply .github/workflows/mac-test.yml:54]: The reason was discussed here https://github.com/gwastro/pycbc/pull/4208 ... It may no longer apply.

**Comment by duncanmmacleod** [comment]
> We list the same `conda-deps` twice now, is it possible to deduplicate that with something like `{[base-conda-deps]}`?

@titodalcanton I have attempted to clean this up by moving the `conda_deps` into `[base]` and creating a new `[bbhx]` section with the options specific to that extra project.

I am not very familiar with tox and tox.ini syntax, so I could easily have done something bad.

**Review by titodalcanton — APPROVED** [review]
As far as I can tell the tests now work for Mac as well, they pass, and the CI configuration has been simplified.

**Comment by spxiwh** [comment]
Thanks @duncanmmacleod ... I see no further comments on this, and the tests are passing, so merging now.
